addon/components/base/bs-form.js

```diff
@@ -220,26 +246,40 @@ export default Component.extend({
    * @private
    */
   submit(e) {
+    this.set('isSubmitting', true);
+
     if (e) {
       e.preventDefault();
     }
     let model = this.get('model');
 
     this.get('onBefore')(model);
 
-    if (!this.get('hasValidator')) {
-      return this.get('onSubmit')(model);
-    } else {
-      let validationPromise = this.validate(this.get('model'));
-      if (validationPromise && validationPromise instanceof RSVP.Promise) {
-        validationPromise
-          .then((r) => this.get('onSubmit')(model, r))
-          .catch((err) => {
-            this.set('showAllValidations', true);
-            return this.get('onInvalid')(model, err);
-          });
-      }
+    let validationPromise = this.get('hasValidator') ? this.validate(this.get('model')) : RSVP.resolve();
+    if (!isPromise(validationPromise)) {
+      validationPromise = RSVP.resolve();
     }
+    validationPromise
+      .then((r) => {
+        let onSubmitPromise = this.get('onSubmit')(model, r);
+        if (!isPromise(onSubmitPromise)) {
+          onSubmitPromise = RSVP.resolve();
+        }
+        onSubmitPromise
+          .finally(() => {
+            if (!this.get('isDestroyed')) {
+              this.set('isSubmitting', false);
+            }
+          });
+      })
+      .catch((err) => {
+        if (!this.get('isDestroyed')) {
+          this.set('showAllValidations', true);
+          this.set('isSubmitting', false);
+        }
+
+        this.get('onInvalid')(model, err);
```

Comment:
Not quite sure if `onInvalid` should also be called if component is already destroyed. To be honest, I'm not sure about the previous behavior. Is it still executed after `this.set()` has thrown cause of setting on a destroyed object?

Reply:
Not sure, but I guess a destroyed thing should never call anything!? 😬